PartitionNotExistException will be thrown when restore from checkpoint

### Search before asking

- [x] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.


### Fluss version

main (development)

### Please describe the bug 🐞

When remove a partition from Fluss, Flink will discovery the removed partition,  and send partition removed event to source reader, source reader will then unsubstribe the partition.. But the partition won't be removed from the state of the soure reader.. When restore, the soure reader will still try to substribe  the partitiion assigned, but since it has been removed from fluss.. It'll then throw `PartitionNotExistException`.. 

Can be reproduced by the following ITCase:
```
class FlinkTableSourceFailOverITCase {

    private static final String CATALOG_NAME = "testcatalog";

    @RegisterExtension
    public static final FlussClusterExtension FLUSS_CLUSTER_EXTENSION =
            FlussClusterExtension.builder()
                    .setClusterConf(
                            new com.alibaba.fluss.config.Configuration()
                                    // set snapshot interval to 1s for testing purposes
                                    .set(ConfigOptions.KV_SNAPSHOT_INTERVAL, Duration.ofSeconds(1))
                                    // not to clean snapshots for test purpose
                                    .set(
                                            ConfigOptions.KV_MAX_RETAINED_SNAPSHOTS,
                                            Integer.MAX_VALUE))
                    .setNumOfTabletServers(3)
                    .build();

    @TempDir File checkpointDir;
    @TempDir File savepointDir;

    protected static Connection conn;
    protected static Admin admin;

    protected static com.alibaba.fluss.config.Configuration clientConf;

    @BeforeAll
    protected static void beforeAll() {
        clientConf = FLUSS_CLUSTER_EXTENSION.getClientConfig();
        conn = ConnectionFactory.createConnection(clientConf);
        admin = conn.getAdmin();
    }

    @Test
    void testRestore() throws Exception {
        final int numTaskManagers = 2;
        final int numSlotsPerTaskManager = 2;
        final int parallelism = numTaskManagers * numSlotsPerTaskManager;

        final MiniClusterResourceFactory clusterFactory =
                new MiniClusterResourceFactory(
                        numTaskManagers,
                        numSlotsPerTaskManager,
                        getFileBasedCheckpointsConfig(savepointDir));

        StreamExecutionEnvironment execEnv = StreamExecutionEnvironment.getExecutionEnvironment();
        execEnv.setParallelism(parallelism);
        StreamTableEnvironment tEnv =
                StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
        String bootstrapServers = String.join(",", clientConf.get(ConfigOptions.BOOTSTRAP_SERVERS));
        // crate catalog using sql
        tEnv.executeSql(
                String.format(
                        "create catalog %s with ('type' = 'fluss-aplus', '%s' = '%s')",
                        CATALOG_NAME, BOOTSTRAP_SERVERS.key(), bootstrapServers));
        tEnv.executeSql("use catalog " + CATALOG_NAME);

        tEnv.executeSql(
                "create table test_partitioned("
                        + "a int, b varchar"
                        + ") partitioned by (b) "
                        + "with ("
                        + "'table.auto-partition.enabled' = 'true',"
                        + "'table.auto-partition.time-unit' = 'year',"
                        + "'scan.partition.discovery.interval' = '100ms',"
                        + "'table.auto-partition.num-precreate' = '1')");

        tEnv.executeSql(
                "create temporary table test_partitioned_sink(a int, b varchar) with ('connector' = 'blackhole')");

        Table table = tEnv.sqlQuery("select * from test_partitioned");
        tEnv.toDataStream(table).addSink(new DiscardingSink<>());

        JobGraph jobGraph = execEnv.getStreamGraph().getJobGraph();

        JobID jobId = jobGraph.getJobID();

        MiniClusterWithClientResource cluster = clusterFactory.get();
        cluster.before();
        ClusterClient<?> client = cluster.getClusterClient();

        String savePointPath;

        try {
            client.submitJob(jobGraph).get();
            waitForAllTaskRunning(cluster.getMiniCluster(), jobId, false);

            // create a partition manually
            FlinkTestBase.createPartitions(
                    FLUSS_CLUSTER_EXTENSION.getZooKeeperClient(),
                    TablePath.of("fluss", "test_partitioned"),
                    Collections.singletonList("2000"));
            // wait from a while to wait the source discovery the partition changes.
            Thread.sleep(3000);

            // drop a partition manually,
            FlinkTestBase.dropPartitions(
                    FLUSS_CLUSTER_EXTENSION.getZooKeeperClient(),
                    TablePath.of("fluss", "test_partitioned"),
                    Collections.singleton("2000"));

            // wait from a while to wait the source unsubscribe the partition
            Thread.sleep(3000);

            // now, stop the job with save point
            savePointPath =
                    client.cancelWithSavepoint(jobId, null, SavepointFormatType.CANONICAL).get();

            execEnv = StreamExecutionEnvironment.getExecutionEnvironment();
            execEnv.setParallelism(parallelism);
            tEnv = StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
            // crate catalog using sql
            tEnv.executeSql(
                    String.format(
                            "create catalog %s with ('type' = 'fluss-aplus', '%s' = '%s')",
                            CATALOG_NAME, BOOTSTRAP_SERVERS.key(), bootstrapServers));
            tEnv.executeSql("use catalog " + CATALOG_NAME);
            table = tEnv.sqlQuery("select * from test_partitioned");
            tEnv.toDataStream(table).addSink(new DiscardingSink<>());
            jobGraph = execEnv.getStreamGraph().getJobGraph();
            SavepointRestoreSettings savepointRestoreSettings =
                    SavepointRestoreSettings.forPath(savePointPath, false, RestoreMode.CLAIM.CLAIM);
            jobGraph.setSavepointRestoreSettings(savepointRestoreSettings);
            client.submitJob(jobGraph).get();
            jobId = jobGraph.getJobID();

            waitForAllTaskRunning(cluster.getMiniCluster(), jobId, false);
        } finally {
            cluster.after();
        }
    }

    private Configuration getFileBasedCheckpointsConfig(File savepointDir) {
        return getFileBasedCheckpointsConfig(savepointDir.toURI().toString());
    }

    private Configuration getFileBasedCheckpointsConfig(final String savepointDir) {
        final Configuration config = new Configuration();
        config.set(StateBackendOptions.STATE_BACKEND, "hashmap");
        config.set(CheckpointingOptions.CHECKPOINTS_DIRECTORY, checkpointDir.toURI().toString());
        config.set(CheckpointingOptions.FS_SMALL_FILE_THRESHOLD, MemorySize.ZERO);
        config.set(CheckpointingOptions.SAVEPOINT_DIRECTORY, savepointDir);
        return config;
    }

    private static class MiniClusterResourceFactory {
        private final int numTaskManagers;
        private final int numSlotsPerTaskManager;
        private final Configuration config;

        private MiniClusterResourceFactory(
                int numTaskManagers, int numSlotsPerTaskManager, Configuration config) {
            this.numTaskManagers = numTaskManagers;
            this.numSlotsPerTaskManager = numSlotsPerTaskManager;
            this.config = config;
        }

        MiniClusterWithClientResource get() {
            return new MiniClusterWithClientResource(
                    new MiniClusterResourceConfiguration.Builder()
                            .setConfiguration(config)
                            .setNumberTaskManagers(numTaskManagers)
                            .setNumberSlotsPerTaskManager(numSlotsPerTaskManager)
                            .build());
        }
    }
}
```

When run the test, we can see the following exception:
```
Caused by: com.alibaba.fluss.exception.FlussRuntimeException: Failed to update metadata
	at com.alibaba.fluss.client.metadata.MetadataUpdater.updateMetadata(MetadataUpdater.java:272) ~[classes/:?]
	at com.alibaba.fluss.client.metadata.MetadataUpdater.checkAndUpdatePartitionMetadata(MetadataUpdater.java:228) ~[classes/:?]
	at com.alibaba.fluss.client.scanner.log.FlussLogScanner.subscribe(FlussLogScanner.java:196) ~[classes/:?]
	at com.alibaba.fluss.connector.flink.source.reader.FlinkSourceSplitReader.subscribeLog(FlinkSourceSplitReader.java:243) ~[classes/:?]
	at com.alibaba.fluss.connector.flink.source.reader.FlinkSourceSplitReader.handleSplitsChanges(FlinkSourceSplitReader.java:188) ~[classes/:?]
	at org.apache.flink.connector.base.source.reader.fetcher.AddSplitsTask.run(AddSplitsTask.java:51) ~[flink-connector-base-1.20.0.jar:1.20.0]
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:165) ~[flink-connector-base-1.20.0.jar:1.20.0]
	... 6 more
```

### Solution

1: when source reader recieve the partition removed event, split source reader should mark it as finished so that it can be removed from split state, so that the removed partition won't be restored to subscribe

2: it may still happens that Fluss drop a partition, but split source reader haven't been aware of it.. After restore, it'll try to subscribe the removed partition which cause the `PartitionNotExistException`.. We may need to warn it instead of throwing exception direclty..

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/enumerator/FlinkSourceEnumerator.java
+++ b/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/enumerator/FlinkSourceEnumerator.java
@@ -124,6 +124,8 @@ public class FlinkSourceEnumerator
 
     private boolean lakeEnabled = false;
 
+    private volatile boolean closed = false;
+
     public FlinkSourceEnumerator(
             TablePath tablePath,
             Configuration flussConf,
@@ -266,6 +268,10 @@ public class FlinkSourceEnumerator
 
     /** Init the splits for Fluss. */
     private void checkPartitionChanges(Set<PartitionInfo> partitionInfos, Throwable t) {
+        if (closed) {
+            // skip if the enumerator is closed to avoid unnecessary error logs
+            return;
+        }
         if (t != null) {
             LOG.error("Failed to list partitions for {}", tablePath, t);
             return;
@@ -678,6 +684,7 @@ public class FlinkSourceEnumerator
     @Override
     public void close() throws IOException {
         try {
+            closed = true;
             if (flussAdmin != null) {
                 flussAdmin.close();
             }

--- a/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/reader/FlinkSourceSplitReader.java
+++ b/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/reader/FlinkSourceSplitReader.java
@@ -104,6 +104,8 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
     private LakeSplitReaderGenerator lakeSplitReaderGenerator;
 
     private final Set<String> emptyLogSplits;
+    // track split IDs corresponding to removed partitions
+    private final Set<String> removedSplits = new HashSet<>();
 
     public FlinkSourceSplitReader(
             Configuration flussConf,
@@ -145,6 +147,12 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
                         new FlinkRecordsWithSplitIds(
                                 new HashSet<>(emptyLogSplits), flinkSourceReaderMetrics);
                 emptyLogSplits.clear();
+                return records;
+            } else if (!removedSplits.isEmpty()) {
+                FlinkRecordsWithSplitIds records =
+                        new FlinkRecordsWithSplitIds(
+                                new HashSet<>(removedSplits), flinkSourceReaderMetrics);
+                removedSplits.clear();
                 return records;
             } else {
                 // if not subscribe any buckets, just return empty records
@@ -250,13 +258,23 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
                     // If no partition id, subscribe by bucket only.
                     logScanner.subscribe(bucket, startingOffset);
                 }
-            } catch (PartitionNotExistException e) {
-                // Instead of throwing, log a warning and return without adding the bucket.
-                LOG.warn(
-                        "Partition {} does not exist when subscribing to log for split {}. Skipping subscription.",
-                        partitionId,
-                        split.splitId());
-                return;
+            } catch (Exception e) {
+                // Traverse the exception chain to check for PartitionNotExistException.
+                Throwable cause = e;
+                boolean partitionNotExist = false;
+                while (cause != null) {
+                    if (cause instanceof PartitionNotExistException) {
+                        partitionNotExist = true;
+                        break;
+                    }
+                    cause = cause.getCause();
+                }
+                if (partitionNotExist) {
+                    LOG.warn(
+                            "Partition {} does not exist when subscribing to log for split {}. Skipping subscription.",
+                            partitionId,
+                            split.splitId());
+                }
             }
 
             LOG.info(
@@ -282,6 +300,7 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
             if (removedPartitions.containsKey(currentBucket.getPartitionId())) {
                 try {
                     // Mark the current split as finished.
+                    removedSplits.add(currentBoundedSplit.splitId());
                     closeCurrentBoundedSplit();
                     unsubscribedTableBuckets.add(currentBucket);
                     LOG.info(
@@ -303,6 +322,7 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
             SourceSplitBase split = snapshotSplitIterator.next();
             TableBucket tableBucket = split.getTableBucket();
             if (removedPartitions.containsKey(tableBucket.getPartitionId())) {
+                removedSplits.add(split.splitId());
                 snapshotSplitIterator.remove();
                 unsubscribedTableBuckets.add(tableBucket);
                 LOG.info(
@@ -323,6 +343,7 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
                 logScanner.unsubscribe(
                         checkNotNull(tableBucket.getPartitionId(), "partition id must be not null"),
                         tableBucket.getBucket());
+                removedSplits.add(tableBucketAndSplit.getValue());
                 subscribeTableBucketIterator.remove();
                 unsubscribedTableBuckets.add(tableBucket);
                 LOG.info(
@@ -472,7 +493,7 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
         return stoppingOffsets.getOrDefault(tableBucket, Long.MAX_VALUE);
     }
 
-    public FlinkRecordsWithSplitIds finishCurrentBoundedSplit() throws IOException {
+    private FlinkRecordsWithSplitIds finishCurrentBoundedSplit() throws IOException {
         Set<String> finishedSplits =
                 currentBoundedSplit instanceof HybridSnapshotLogSplit
                         // is hybrid split, not to finish this split

--- a/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/reader/FlinkSourceSplitReader.java
+++ b/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/reader/FlinkSourceSplitReader.java
@@ -36,6 +36,7 @@ import com.alibaba.fluss.metadata.TableBucket;
 import com.alibaba.fluss.metadata.TablePath;
 import com.alibaba.fluss.types.RowType;
 import com.alibaba.fluss.utils.CloseableIterator;
+import com.alibaba.fluss.utils.ExceptionUtils;
 
 import org.apache.flink.connector.base.source.reader.RecordsWithSplitIds;
 import org.apache.flink.connector.base.source.reader.splitreader.SplitReader;
@@ -260,20 +261,15 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
                 }
             } catch (Exception e) {
                 // Traverse the exception chain to check for PartitionNotExistException.
-                Throwable cause = e;
-                boolean partitionNotExist = false;
-                while (cause != null) {
-                    if (cause instanceof PartitionNotExistException) {
-                        partitionNotExist = true;
-                        break;
-                    }
-                    cause = cause.getCause();
-                }
+                boolean partitionNotExist =
+                        ExceptionUtils.findThrowable(e, PartitionNotExistException.class)
+                                .isPresent();
                 if (partitionNotExist) {
                     LOG.warn(
                             "Partition {} does not exist when subscribing to log for split {}. Skipping subscription.",
                             partitionId,
                             split.splitId());
+                    return;
                 }
             }
 

--- a/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/reader/FlinkSourceSplitReader.java
+++ b/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/reader/FlinkSourceSplitReader.java
@@ -31,6 +31,7 @@ import com.alibaba.fluss.connector.flink.source.split.HybridSnapshotLogSplit;
 import com.alibaba.fluss.connector.flink.source.split.LogSplit;
 import com.alibaba.fluss.connector.flink.source.split.SnapshotSplit;
 import com.alibaba.fluss.connector.flink.source.split.SourceSplitBase;
+import com.alibaba.fluss.exception.PartitionNotExistException;
 import com.alibaba.fluss.metadata.TableBucket;
 import com.alibaba.fluss.metadata.TablePath;
 import com.alibaba.fluss.types.RowType;
@@ -241,18 +242,28 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
         } else {
             Long partitionId = tableBucket.getPartitionId();
             int bucket = tableBucket.getBucket();
-            if (partitionId != null) {
-                logScanner.subscribe(partitionId, bucket, startingOffset);
-            } else {
-                logScanner.subscribe(bucket, startingOffset);
+            try {
+                if (partitionId != null) {
+                    // Try to subscribe using the partition id.
+                    logScanner.subscribe(partitionId, bucket, startingOffset);
+                } else {
+                    // If no partition id, subscribe by bucket only.
+                    logScanner.subscribe(bucket, startingOffset);
+                }
+            } catch (PartitionNotExistException e) {
+                // Instead of throwing, log a warning and return without adding the bucket.
+                LOG.warn(
+                        "Partition {} does not exist when subscribing to log for split {}. Skipping subscription.",
+                        partitionId,
+                        split.splitId());
+                return;
             }
 
             LOG.info(
                     "Subscribe to read log for split {} from offset {}.",
                     split.splitId(),
                     startingOffset);
-
-            // Track the new bucket in metrics
+            // Track the new bucket in metrics and internal state.
             flinkSourceReaderMetrics.registerTableBucket(tableBucket);
             subscribedBuckets.put(tableBucket, split.splitId());
         }
@@ -262,8 +273,31 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
         // todo, may consider to close the current snapshot reader if
         // the current snapshot split is in the partition buckets
 
-        // may remove from pending snapshot splits
+        // Set to collect table buckets that are unsubscribed.
         Set<TableBucket> unsubscribedTableBuckets = new HashSet<>();
+        // First, if the current active bounded split belongs to a removed partition,
+        // finish it so it will not be restored.
+        if (currentBoundedSplit != null) {
+            TableBucket currentBucket = currentBoundedSplit.getTableBucket();
+            if (removedPartitions.containsKey(currentBucket.getPartitionId())) {
+                try {
+                    // Mark the current split as finished.
+                    closeCurrentBoundedSplit();
+                    unsubscribedTableBuckets.add(currentBucket);
+                    LOG.info(
+                            "Mark current bounded split {} as finished for removed partition {}.",
+                            currentBucket,
+                            removedPartitions.get(currentBucket.getPartitionId()));
+                } catch (IOException e) {
+                    LOG.warn(
+                            "Failed to close current bounded split for removed partition {}.",
+                            currentBucket.getPartitionId(),
+                            e);
+                }
+            }
+        }
+
+        // Remove pending snapshot splits whose table buckets belong to removed partitions.
         Iterator<SourceSplitBase> snapshotSplitIterator = boundedSplits.iterator();
         while (snapshotSplitIterator.hasNext()) {
             SourceSplitBase split = snapshotSplitIterator.next();
@@ -272,7 +306,7 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
                 snapshotSplitIterator.remove();
                 unsubscribedTableBuckets.add(tableBucket);
                 LOG.info(
-                        "Cancel to read snapshot split {} for non-existed partition {}.",
+                        "Cancel reading snapshot split {} for removed partition {}.",
                         split.splitId(),
                         removedPartitions.get(tableBucket.getPartitionId()));
             }

--- a/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/reader/FlinkSourceSplitReader.java
+++ b/fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/reader/FlinkSourceSplitReader.java
@@ -132,6 +132,13 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
 
     @Override
     public RecordsWithSplitIds<RecordAndPos> fetch() throws IOException {
+        if (!removedSplits.isEmpty()) {
+            FlinkRecordsWithSplitIds records =
+                    new FlinkRecordsWithSplitIds(
+                            new HashSet<>(removedSplits), flinkSourceReaderMetrics);
+            removedSplits.clear();
+            return records;
+        }
         checkSnapshotSplitOrStartNext();
         if (currentBoundedSplitReader != null) {
             CloseableIterator<RecordAndPos> recordIterator = currentBoundedSplitReader.readBatch();
@@ -148,12 +155,6 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
                         new FlinkRecordsWithSplitIds(
                                 new HashSet<>(emptyLogSplits), flinkSourceReaderMetrics);
                 emptyLogSplits.clear();
-                return records;
-            } else if (!removedSplits.isEmpty()) {
-                FlinkRecordsWithSplitIds records =
-                        new FlinkRecordsWithSplitIds(
-                                new HashSet<>(removedSplits), flinkSourceReaderMetrics);
-                removedSplits.clear();
                 return records;
             } else {
                 // if not subscribe any buckets, just return empty records
@@ -251,26 +252,30 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
         } else {
             Long partitionId = tableBucket.getPartitionId();
             int bucket = tableBucket.getBucket();
-            try {
-                if (partitionId != null) {
-                    // Try to subscribe using the partition id.
+            if (partitionId != null) {
+                // Try to subscribe using the partition id.
+                try {
                     logScanner.subscribe(partitionId, bucket, startingOffset);
-                } else {
-                    // If no partition id, subscribe by bucket only.
-                    logScanner.subscribe(bucket, startingOffset);
+                } catch (Exception e) {
+                    // the PartitionNotExistException may still happens when partition is removed
+                    // but Flink source reader failover before aware of it
+                    // Traverse the exception chain to check for PartitionNotExistException.
+                    boolean partitionNotExist =
+                            ExceptionUtils.findThrowable(e, PartitionNotExistException.class)
+                                    .isPresent();
+                    if (partitionNotExist) {
+                        // mark the not exist partition to be removed
+                        removedSplits.add(split.splitId());
+                        LOG.warn(
+                                "Partition {} does not exist when subscribing to log for split {}. Skipping subscription.",
+                                partitionId,
+                                split.splitId());
+                        return;
+                    }
                 }
-            } catch (Exception e) {
-                // Traverse the exception chain to check for PartitionNotExistException.
-                boolean partitionNotExist =
-                        ExceptionUtils.findThrowable(e, PartitionNotExistException.class)
-                                .isPresent();
-                if (partitionNotExist) {
-                    LOG.warn(
-                            "Partition {} does not exist when subscribing to log for split {}. Skipping subscription.",
-                            partitionId,
-                            split.splitId());
-                    return;
-                }
+            } else {
+                // If no partition id, subscribe by bucket only.
+                logScanner.subscribe(bucket, startingOffset);
             }
 
             LOG.info(
@@ -284,9 +289,6 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
     }
 
     public Set<TableBucket> removePartitions(Map<Long, String> removedPartitions) {
-        // todo, may consider to close the current snapshot reader if
-        // the current snapshot split is in the partition buckets
-
         // Set to collect table buckets that are unsubscribed.
         Set<TableBucket> unsubscribedTableBuckets = new HashSet<>();
         // First, if the current active bounded split belongs to a removed partition,
@@ -306,7 +308,7 @@ public class FlinkSourceSplitReader implements SplitReader<RecordAndPos, SourceS
                 } catch (IOException e) {
                     LOG.warn(
                             "Failed to close current bounded split for removed partition {}.",
-                            currentBucket.getPartitionId(),
+                            removedPartitions.get(currentBucket.getPartitionId()),
                             e);
                 }
             }

--- a/fluss-connectors/fluss-connector-flink/src/test/java/com/alibaba/fluss/connector/flink/source/FlinkTableSourceFailOverITCase.java
+++ b/fluss-connectors/fluss-connector-flink/src/test/java/com/alibaba/fluss/connector/flink/source/FlinkTableSourceFailOverITCase.java
@@ -1,0 +1,223 @@
+/*
+ *  Copyright (c) 2024 Alibaba Group Holding Ltd.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.alibaba.fluss.connector.flink.source;
+
+import com.alibaba.fluss.client.Connection;
+import com.alibaba.fluss.client.ConnectionFactory;
+import com.alibaba.fluss.client.admin.Admin;
+import com.alibaba.fluss.config.ConfigOptions;
+import com.alibaba.fluss.connector.flink.source.testutils.FlinkTestBase;
+import com.alibaba.fluss.metadata.TablePath;
+import com.alibaba.fluss.server.testutils.FlussClusterExtension;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.client.program.ClusterClient;
+import org.apache.flink.configuration.CheckpointingOptions;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.MemorySize;
+import org.apache.flink.configuration.StateBackendOptions;
+import org.apache.flink.core.execution.RestoreMode;
+import org.apache.flink.core.execution.SavepointFormatType;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.SavepointRestoreSettings;
+import org.apache.flink.runtime.testutils.MiniClusterResourceConfiguration;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
+import org.apache.flink.table.api.EnvironmentSettings;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.test.util.MiniClusterWithClientResource;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.File;
+import java.time.Duration;
+import java.util.Collections;
+
+import static com.alibaba.fluss.connector.flink.FlinkConnectorOptions.BOOTSTRAP_SERVERS;
+import static org.apache.flink.runtime.testutils.CommonTestUtils.waitForAllTaskRunning;
+
+/** IT case for flink table source fail over. */
+class FlinkTableSourceFailOverITCase {
+
+    private static final String CATALOG_NAME = "testcatalog";
+
+    @RegisterExtension
+    public static final FlussClusterExtension FLUSS_CLUSTER_EXTENSION =
+            FlussClusterExtension.builder()
+                    .setClusterConf(
+                            new com.alibaba.fluss.config.Configuration()
+                                    // set snapshot interval to 1s for testing purposes
+                                    .set(ConfigOptions.KV_SNAPSHOT_INTERVAL, Duration.ofSeconds(1))
+                                    // not to clean snapshots for test purpose
+                                    .set(
+                                            ConfigOptions.KV_MAX_RETAINED_SNAPSHOTS,
+                                            Integer.MAX_VALUE))
+                    .setNumOfTabletServers(3)
+                    .build();
+
+    @TempDir File checkpointDir;
+    @TempDir File savepointDir;
+
+    protected static Connection conn;
+    protected static Admin admin;
+
+    protected static com.alibaba.fluss.config.Configuration clientConf;
+
+    @BeforeAll
+    protected static void beforeAll() {
+        clientConf = FLUSS_CLUSTER_EXTENSION.getClientConfig();
+        conn = ConnectionFactory.createConnection(clientConf);
+        admin = conn.getAdmin();
+    }
+
+    @Test
+    void testRestore() throws Exception {
+        final int numTaskManagers = 2;
+        final int numSlotsPerTaskManager = 2;
+        final int parallelism = numTaskManagers * numSlotsPerTaskManager;
+
+        final MiniClusterResourceFactory clusterFactory =
+                new MiniClusterResourceFactory(
+                        numTaskManagers,
+                        numSlotsPerTaskManager,
+                        getFileBasedCheckpointsConfig(savepointDir));
+
+        StreamExecutionEnvironment execEnv = StreamExecutionEnvironment.getExecutionEnvironment();
+        execEnv.setParallelism(parallelism);
+        StreamTableEnvironment tEnv =
+                StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
+        String bootstrapServers = String.join(",", clientConf.get(ConfigOptions.BOOTSTRAP_SERVERS));
+        // crate catalog using sql
+        tEnv.executeSql(
+                String.format(
+                        "create catalog %s with ('type' = 'fluss', '%s' = '%s')",
+                        CATALOG_NAME, BOOTSTRAP_SERVERS.key(), bootstrapServers));
+        tEnv.executeSql("use catalog " + CATALOG_NAME);
+
+        tEnv.executeSql(
+                "create table test_partitioned("
+                        + "a int, b varchar"
+                        + ") partitioned by (b) "
+                        + "with ("
+                        + "'table.auto-partition.enabled' = 'true',"
+                        + "'table.auto-partition.time-unit' = 'year',"
+                        + "'scan.partition.discovery.interval' = '100ms',"
+                        + "'table.auto-partition.num-precreate' = '1')");
+
+        tEnv.executeSql(
+                "create temporary table test_partitioned_sink(a int, b varchar) with ('connector' = 'blackhole')");
+
+        Table table = tEnv.sqlQuery("select * from test_partitioned");
+        tEnv.toDataStream(table).addSink(new DiscardingSink<>());
+
+        JobGraph jobGraph = execEnv.getStreamGraph().getJobGraph();
+
+        JobID jobId = jobGraph.getJobID();
+
+        MiniClusterWithClientResource cluster = clusterFactory.get();
+        cluster.before();
+        ClusterClient<?> client = cluster.getClusterClient();
+
+        String savePointPath;
+
+        try {
+            client.submitJob(jobGraph).get();
+            waitForAllTaskRunning(cluster.getMiniCluster(), jobId, false);
+
+            // create a partition manually
+            FlinkTestBase.createPartitions(
+                    FLUSS_CLUSTER_EXTENSION.getZooKeeperClient(),
+                    TablePath.of("fluss", "test_partitioned"),
+                    Collections.singletonList("2000"));
+            // wait from a while to wait the source discovery the partition changes.
+            Thread.sleep(3000);
+
+            // drop a partition manually,
+            FlinkTestBase.dropPartitions(
+                    FLUSS_CLUSTER_EXTENSION.getZooKeeperClient(),
+                    TablePath.of("fluss", "test_partitioned"),
+                    Collections.singleton("2000"));
+
+            // wait from a while to wait the source unsubscribe the partition
+            Thread.sleep(3000);
+
+            // now, stop the job with save point
+            savePointPath =
+                    client.cancelWithSavepoint(jobId, null, SavepointFormatType.CANONICAL).get();
+
+            execEnv = StreamExecutionEnvironment.getExecutionEnvironment();
+            execEnv.setParallelism(parallelism);
+            tEnv = StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
+            // crate catalog using sql
+            tEnv.executeSql(
+                    String.format(
+                            "create catalog %s with ('type' = 'fluss', '%s' = '%s')",
+                            CATALOG_NAME, BOOTSTRAP_SERVERS.key(), bootstrapServers));
+            tEnv.executeSql("use catalog " + CATALOG_NAME);
+            table = tEnv.sqlQuery("select * from test_partitioned");
+            tEnv.toDataStream(table).addSink(new DiscardingSink<>());
+            jobGraph = execEnv.getStreamGraph().getJobGraph();
+            SavepointRestoreSettings savepointRestoreSettings =
+                    SavepointRestoreSettings.forPath(savePointPath, false, RestoreMode.CLAIM.CLAIM);
+            jobGraph.setSavepointRestoreSettings(savepointRestoreSettings);
+            client.submitJob(jobGraph).get();
+            jobId = jobGraph.getJobID();
+
+            waitForAllTaskRunning(cluster.getMiniCluster(), jobId, false);
+        } finally {
+            cluster.after();
+        }
+    }
+
+    private Configuration getFileBasedCheckpointsConfig(File savepointDir) {
+        return getFileBasedCheckpointsConfig(savepointDir.toURI().toString());
+    }
+
+    private Configuration getFileBasedCheckpointsConfig(final String savepointDir) {
+        final Configuration config = new Configuration();
+        config.set(StateBackendOptions.STATE_BACKEND, "hashmap");
+        config.set(CheckpointingOptions.CHECKPOINTS_DIRECTORY, checkpointDir.toURI().toString());
+        config.set(CheckpointingOptions.FS_SMALL_FILE_THRESHOLD, MemorySize.ZERO);
+        config.set(CheckpointingOptions.SAVEPOINT_DIRECTORY, savepointDir);
+        return config;
+    }
+
+    private static class MiniClusterResourceFactory {
+        private final int numTaskManagers;
+        private final int numSlotsPerTaskManager;
+        private final Configuration config;
+
+        private MiniClusterResourceFactory(
+                int numTaskManagers, int numSlotsPerTaskManager, Configuration config) {
+            this.numTaskManagers = numTaskManagers;
+            this.numSlotsPerTaskManager = numSlotsPerTaskManager;
+            this.config = config;
+        }
+
+        MiniClusterWithClientResource get() {
+            return new MiniClusterWithClientResource(
+                    new MiniClusterResourceConfiguration.Builder()
+                            .setConfiguration(config)
+                            .setNumberTaskManagers(numTaskManagers)
+                            .setNumberSlotsPerTaskManager(numSlotsPerTaskManager)
+                            .build());
+        }
+    }
+}

--- a/fluss-connectors/fluss-connector-flink/src/test/java/com/alibaba/fluss/connector/flink/source/FlinkTableSourceFailOverITCase.java
+++ b/fluss-connectors/fluss-connector-flink/src/test/java/com/alibaba/fluss/connector/flink/source/FlinkTableSourceFailOverITCase.java
@@ -18,48 +18,54 @@ package com.alibaba.fluss.connector.flink.source;
 
 import com.alibaba.fluss.client.Connection;
 import com.alibaba.fluss.client.ConnectionFactory;
-import com.alibaba.fluss.client.admin.Admin;
+import com.alibaba.fluss.client.table.Table;
+import com.alibaba.fluss.client.table.writer.AppendWriter;
 import com.alibaba.fluss.config.ConfigOptions;
-import com.alibaba.fluss.connector.flink.source.testutils.FlinkTestBase;
-import com.alibaba.fluss.exception.PartitionNotExistException;
 import com.alibaba.fluss.metadata.TablePath;
 import com.alibaba.fluss.server.testutils.FlussClusterExtension;
+import com.alibaba.fluss.server.zk.ZooKeeperClient;
 
-import org.apache.flink.api.common.JobID;
-import org.apache.flink.client.program.ClusterClient;
 import org.apache.flink.configuration.CheckpointingOptions;
 import org.apache.flink.configuration.Configuration;
 import org.apache.flink.configuration.MemorySize;
 import org.apache.flink.configuration.StateBackendOptions;
-import org.apache.flink.core.execution.RestoreMode;
+import org.apache.flink.configuration.StateRecoveryOptions;
 import org.apache.flink.core.execution.SavepointFormatType;
-import org.apache.flink.runtime.jobgraph.JobGraph;
-import org.apache.flink.runtime.jobgraph.SavepointRestoreSettings;
-import org.apache.flink.runtime.jobmaster.JobResult;
 import org.apache.flink.runtime.testutils.MiniClusterResourceConfiguration;
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
-import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
 import org.apache.flink.table.api.EnvironmentSettings;
-import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableResult;
 import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
 import org.apache.flink.test.util.MiniClusterWithClientResource;
-import org.junit.jupiter.api.BeforeAll;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.CloseableIterator;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 import org.junit.jupiter.api.io.TempDir;
 
 import java.io.File;
 import java.time.Duration;
+import java.time.Year;
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.List;
 
 import static com.alibaba.fluss.connector.flink.FlinkConnectorOptions.BOOTSTRAP_SERVERS;
-import static org.apache.flink.runtime.testutils.CommonTestUtils.waitForAllTaskRunning;
-import static org.assertj.core.api.Assertions.assertThat;
+import static com.alibaba.fluss.connector.flink.source.testutils.FlinkTestBase.assertResultsIgnoreOrder;
+import static com.alibaba.fluss.connector.flink.source.testutils.FlinkTestBase.createPartitions;
+import static com.alibaba.fluss.connector.flink.source.testutils.FlinkTestBase.dropPartitions;
+import static com.alibaba.fluss.connector.flink.source.testutils.FlinkTestBase.waitUntilPartitions;
+import static com.alibaba.fluss.testutils.DataTestUtils.row;
 
 /** IT case for flink table source fail over. */
 class FlinkTableSourceFailOverITCase {
 
     private static final String CATALOG_NAME = "testcatalog";
+
+    @TempDir public static File checkpointDir;
+    @TempDir public static File savepointDir;
 
     @RegisterExtension
     public static final FlussClusterExtension FLUSS_CLUSTER_EXTENSION =
@@ -75,35 +81,25 @@ class FlinkTableSourceFailOverITCase {
                     .setNumOfTabletServers(3)
                     .build();
 
-    @TempDir File checkpointDir;
-    @TempDir File savepointDir;
+    com.alibaba.fluss.config.Configuration clientConf;
+    ZooKeeperClient zkClient;
+    Connection conn;
 
-    protected static Connection conn;
-    protected static Admin admin;
-
-    protected static com.alibaba.fluss.config.Configuration clientConf;
-
-    @BeforeAll
-    protected static void beforeAll() {
+    @BeforeEach
+    protected void beforeEach() {
         clientConf = FLUSS_CLUSTER_EXTENSION.getClientConfig();
+        zkClient = FLUSS_CLUSTER_EXTENSION.getZooKeeperClient();
         conn = ConnectionFactory.createConnection(clientConf);
-        admin = conn.getAdmin();
     }
 
-    @Test
-    void testRestore() throws Exception {
-        final int numTaskManagers = 2;
-        final int numSlotsPerTaskManager = 2;
-        final int parallelism = numTaskManagers * numSlotsPerTaskManager;
+    @AfterEach
+    protected void afterEach() throws Exception {
+        conn.close();
+    }
 
-        final MiniClusterResourceFactory clusterFactory =
-                new MiniClusterResourceFactory(
-                        numTaskManagers,
-                        numSlotsPerTaskManager,
-                        getFileBasedCheckpointsConfig(savepointDir));
-
+    private StreamTableEnvironment initTableEnvironment() {
         StreamExecutionEnvironment execEnv = StreamExecutionEnvironment.getExecutionEnvironment();
-        execEnv.setParallelism(parallelism);
+        execEnv.setParallelism(1);
         StreamTableEnvironment tEnv =
                 StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
         String bootstrapServers = String.join(",", clientConf.get(ConfigOptions.BOOTSTRAP_SERVERS));
@@ -113,123 +109,117 @@ class FlinkTableSourceFailOverITCase {
                         "create catalog %s with ('type' = 'fluss', '%s' = '%s')",
                         CATALOG_NAME, BOOTSTRAP_SERVERS.key(), bootstrapServers));
         tEnv.executeSql("use catalog " + CATALOG_NAME);
+        return tEnv;
+    }
 
-        tEnv.executeSql(
-                "create table test_partitioned("
-                        + "a int, b varchar"
-                        + ") partitioned by (b) "
-                        + "with ("
-                        + "'table.auto-partition.enabled' = 'true',"
-                        + "'table.auto-partition.time-unit' = 'year',"
-                        + "'scan.partition.discovery.interval' = '100ms',"
-                        + "'table.auto-partition.num-precreate' = '1')");
+    @Test
+    void testRestore() throws Exception {
+        final int numTaskManagers = 2;
+        final int numSlotsPerTaskManager = 2;
 
-        tEnv.executeSql(
-                "create temporary table test_partitioned_sink(a int, b varchar) with ('connector' = 'blackhole')");
+        // Start Flink
+        MiniClusterWithClientResource cluster =
+                new MiniClusterWithClientResource(
+                        new MiniClusterResourceConfiguration.Builder()
+                                .setConfiguration(getFileBasedCheckpointsConfig(savepointDir))
+                                .setNumberTaskManagers(numTaskManagers)
+                                .setNumberSlotsPerTaskManager(numSlotsPerTaskManager)
+                                .build());
 
-        Table table = tEnv.sqlQuery("select * from test_partitioned");
-        tEnv.toDataStream(table).addSink(new DiscardingSink<>());
-
-        JobGraph jobGraph = execEnv.getStreamGraph().getJobGraph();
-
-        JobID jobId = jobGraph.getJobID();
-
-        MiniClusterWithClientResource cluster = clusterFactory.get();
         cluster.before();
-        ClusterClient<?> client = cluster.getClusterClient();
-
-        String savePointPath;
 
         try {
-            client.submitJob(jobGraph).get();
-            waitForAllTaskRunning(cluster.getMiniCluster(), jobId, false);
+            StreamTableEnvironment tEnv = initTableEnvironment();
+            tEnv.executeSql(
+                    "create table test_partitioned ("
+                            + "a int, b varchar"
+                            + ") partitioned by (b) "
+                            + "with ("
+                            + "'table.auto-partition.enabled' = 'true',"
+                            + "'table.auto-partition.time-unit' = 'year',"
+                            + "'scan.partition.discovery.interval' = '100ms',"
+                            + "'table.auto-partition.num-precreate' = '1')");
+            tEnv.executeSql("create table result_table (a int, b varchar)");
+
+            TablePath tablePath = TablePath.of("fluss", "test_partitioned");
 
             // create a partition manually
-            FlinkTestBase.createPartitions(
-                    FLUSS_CLUSTER_EXTENSION.getZooKeeperClient(),
-                    TablePath.of("fluss", "test_partitioned"),
-                    Collections.singletonList("2000"));
-            // wait from a while to wait the source discovery the partition changes.
-            Thread.sleep(3000);
+            createPartitions(zkClient, tablePath, Collections.singletonList("4000"));
+            waitUntilPartitions(zkClient, tablePath, 2);
 
-            // drop a partition manually,
-            FlinkTestBase.dropPartitions(
-                    FLUSS_CLUSTER_EXTENSION.getZooKeeperClient(),
-                    TablePath.of("fluss", "test_partitioned"),
-                    Collections.singleton("2000"));
+            // append 3 records for each partition
+            Table table = conn.getTable(tablePath);
+            AppendWriter writer = table.newAppend().createWriter();
+            String thisYear = String.valueOf(Year.now().getValue());
+            List<String> expected = new ArrayList<>();
+            for (int i = 0; i < 3; i++) {
+                writer.append(row(i, thisYear));
+                writer.append(row(i, "4000"));
+                expected.add("+I[" + i + ", " + thisYear + "]");
+                expected.add("+I[" + i + ", 4000]");
+            }
+            writer.flush();
 
-            // wait from a while to wait the source unsubscribe the partition
-            Thread.sleep(3000);
+            // execute the query to fetch logs from the table
+            TableResult insertResult =
+                    tEnv.executeSql("insert into result_table select * from test_partitioned");
+            // we have to create a intermediate table to collect result,
+            // because CollectSink can't be restored from savepoint
+            CloseableIterator<Row> results =
+                    tEnv.executeSql("select * from result_table").collect();
+            assertResultsIgnoreOrder(results, expected, false);
+            expected.clear();
+
+            // drop the partition manually
+            dropPartitions(zkClient, tablePath, Collections.singleton("4000"));
+            waitUntilPartitions(zkClient, tablePath, 1);
+
+            // create a new partition again and append records into it
+            createPartitions(zkClient, tablePath, Collections.singletonList("5000"));
+            waitUntilPartitions(zkClient, tablePath, 2);
+            writer.append(row(4, "5000")).get();
+            expected.add("+I[4, 5000]");
+            // if the source subscribes the new partition successfully,
+            // it should have removed the old partition successfully
+            assertResultsIgnoreOrder(results, expected, false);
+            expected.clear();
 
             // now, stop the job with save point
-            savePointPath =
-                    client.cancelWithSavepoint(jobId, null, SavepointFormatType.CANONICAL).get();
+            String savepointPath =
+                    insertResult
+                            .getJobClient()
+                            .get()
+                            .stopWithSavepoint(
+                                    false,
+                                    savepointDir.getAbsolutePath(),
+                                    SavepointFormatType.CANONICAL)
+                            .get();
 
-            execEnv = StreamExecutionEnvironment.getExecutionEnvironment();
-            execEnv.setParallelism(parallelism);
-            tEnv = StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
-            // crate catalog using sql
-            tEnv.executeSql(
-                    String.format(
-                            "create catalog %s with ('type' = 'fluss', '%s' = '%s')",
-                            CATALOG_NAME, BOOTSTRAP_SERVERS.key(), bootstrapServers));
-            tEnv.executeSql("use catalog " + CATALOG_NAME);
-            table = tEnv.sqlQuery("select * from test_partitioned");
-            tEnv.toDataStream(table).addSink(new DiscardingSink<>());
-            jobGraph = execEnv.getStreamGraph().getJobGraph();
-            SavepointRestoreSettings savepointRestoreSettings =
-                    SavepointRestoreSettings.forPath(savePointPath, false, RestoreMode.CLAIM.CLAIM);
-            jobGraph.setSavepointRestoreSettings(savepointRestoreSettings);
-            client.submitJob(jobGraph).get();
-            jobId = jobGraph.getJobID();
-
-            waitForAllTaskRunning(cluster.getMiniCluster(), jobId, false);
-            JobResult jobResult = client.requestJobResult(jobId).get();
-            if (jobResult.getSerializedThrowable().isPresent()) {
-                assertThat(
-                                jobResult.getSerializedThrowable().get().getCause()
-                                        instanceof PartitionNotExistException)
-                        .isEqualTo(false);
-                System.out.println(
-                        jobResult.getSerializedThrowable().get().getFullStringifiedStackTrace());
-            }
+            tEnv.getConfig().set(StateRecoveryOptions.SAVEPOINT_PATH, savepointPath);
+            insertResult =
+                    tEnv.executeSql("insert into result_table select * from test_partitioned");
+            // append a new row again to check if the source can restore the state correctly
+            writer.append(row(5, "5000")).get();
+            expected.add("+I[5, 5000]");
+            assertResultsIgnoreOrder(results, expected, true);
+            // cancel the insert job
+            insertResult.getJobClient().get().cancel().get();
         } finally {
+            // stop the cluster and thereby cancel the job
             cluster.after();
         }
     }
 
-    private Configuration getFileBasedCheckpointsConfig(File savepointDir) {
+    private static Configuration getFileBasedCheckpointsConfig(File savepointDir) {
         return getFileBasedCheckpointsConfig(savepointDir.toURI().toString());
     }
 
-    private Configuration getFileBasedCheckpointsConfig(final String savepointDir) {
+    private static Configuration getFileBasedCheckpointsConfig(final String savepointDir) {
         final Configuration config = new Configuration();
         config.set(StateBackendOptions.STATE_BACKEND, "hashmap");
         config.set(CheckpointingOptions.CHECKPOINTS_DIRECTORY, checkpointDir.toURI().toString());
         config.set(CheckpointingOptions.FS_SMALL_FILE_THRESHOLD, MemorySize.ZERO);
         config.set(CheckpointingOptions.SAVEPOINT_DIRECTORY, savepointDir);
         return config;
-    }
-
-    private static class MiniClusterResourceFactory {
-        private final int numTaskManagers;
-        private final int numSlotsPerTaskManager;
-        private final Configuration config;
-
-        private MiniClusterResourceFactory(
-                int numTaskManagers, int numSlotsPerTaskManager, Configuration config) {
-            this.numTaskManagers = numTaskManagers;
-            this.numSlotsPerTaskManager = numSlotsPerTaskManager;
-            this.config = config;
-        }
-
-        MiniClusterWithClientResource get() {
-            return new MiniClusterWithClientResource(
-                    new MiniClusterResourceConfiguration.Builder()
-                            .setConfiguration(config)
-                            .setNumberTaskManagers(numTaskManagers)
-                            .setNumberSlotsPerTaskManager(numSlotsPerTaskManager)
-                            .build());
-        }
     }
 }

--- a/fluss-connectors/fluss-connector-flink/src/test/java/com/alibaba/fluss/connector/flink/source/FlinkTableSourceFailOverITCase.java
+++ b/fluss-connectors/fluss-connector-flink/src/test/java/com/alibaba/fluss/connector/flink/source/FlinkTableSourceFailOverITCase.java
@@ -21,6 +21,7 @@ import com.alibaba.fluss.client.ConnectionFactory;
 import com.alibaba.fluss.client.admin.Admin;
 import com.alibaba.fluss.config.ConfigOptions;
 import com.alibaba.fluss.connector.flink.source.testutils.FlinkTestBase;
+import com.alibaba.fluss.exception.PartitionNotExistException;
 import com.alibaba.fluss.metadata.TablePath;
 import com.alibaba.fluss.server.testutils.FlussClusterExtension;
 
@@ -34,6 +35,7 @@ import org.apache.flink.core.execution.RestoreMode;
 import org.apache.flink.core.execution.SavepointFormatType;
 import org.apache.flink.runtime.jobgraph.JobGraph;
 import org.apache.flink.runtime.jobgraph.SavepointRestoreSettings;
+import org.apache.flink.runtime.jobmaster.JobResult;
 import org.apache.flink.runtime.testutils.MiniClusterResourceConfiguration;
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
 import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
@@ -52,6 +54,7 @@ import java.util.Collections;
 
 import static com.alibaba.fluss.connector.flink.FlinkConnectorOptions.BOOTSTRAP_SERVERS;
 import static org.apache.flink.runtime.testutils.CommonTestUtils.waitForAllTaskRunning;
+import static org.assertj.core.api.Assertions.assertThat;
 
 /** IT case for flink table source fail over. */
 class FlinkTableSourceFailOverITCase {
@@ -181,6 +184,15 @@ class FlinkTableSourceFailOverITCase {
             jobId = jobGraph.getJobID();
 
             waitForAllTaskRunning(cluster.getMiniCluster(), jobId, false);
+            JobResult jobResult = client.requestJobResult(jobId).get();
+            if (jobResult.getSerializedThrowable().isPresent()) {
+                assertThat(
+                                jobResult.getSerializedThrowable().get().getCause()
+                                        instanceof PartitionNotExistException)
+                        .isEqualTo(false);
+                System.out.println(
+                        jobResult.getSerializedThrowable().get().getFullStringifiedStackTrace());
+            }
         } finally {
             cluster.after();
         }

